### PR TITLE
Leverage UI

### PR DESCRIPTION
**Pull request by Pawel-Szydlo**

### CODE REVIEW

**Review by wojciech-cichocki — COMMENTED** [review]
Black screen when i select leverage tab with unconnected wallet

File: src/components/Leverage/CloseLeverage/CloseLeverage.tsx [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            classes={{
+              rail: classes.sliderRail,
+              track: classes.sliderTrack,
+              thumb: classes.sliderThumb,
+              markActive: classes.markActive,
+              mark: classes.markActive,
+              markLabel: classes.markLabel,
+              markLabelActive: classes.markLabel
+            }}
+            // onChange={(_event, newValue) => {
```
Comment (awojciak) [inline src/components/Leverage/CloseLeverage/CloseLeverage.tsx:90]: remove this if not used later

File: src/components/Leverage/LeverageAction/LeverageAction.tsx [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  minCRatio: number
+  changeCustomCRatio: (value: string) => void
+  currentLeverage: string
+  maxLeverage: string
+  setLeverageStatus: (value: boolean) => void
+  leverageType: string
+  blockSubmitButton: boolean
+  action: string
+  setAction: (value: string) => void
+  openCloseLeverage: (event: any) => void
```
Comment (awojciak) [inline src/components/Leverage/LeverageAction/LeverageAction.tsx:24]: don't use "any" type

File: src/components/Leverage/LeverageOption/style.ts [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    background: colors.navy.darkGrey
+  },
+  actionButton: {
+    padding: '5px 0',
+    minWidth: '116px',
+    ...typography.subtitle1,
+    marginLeft: '16px',
+    textTransform: 'capitalize',
+    '&:hover': {
+      background: '#4ADFBA'
```
Comment (awojciak) [inline src/components/Leverage/LeverageOption/style.ts:114]: use colors from theme if they are there; comment to thole PR

File: src/store/consts/utils.ts [hunk 4]
```diff
@@ -253,6 +254,27 @@ export const calculateTimeRemaining = (slot: number, nextRoundStartSlot: BN): BN
   }
   return slotDiff.muln(slotTime)
 }
+export const getAddressFromIndex = (
```
Comment (awojciak) [inline src/store/consts/utils.ts:257]: rename this function to 'getSyntheticAndCollateralAdresses', current name isn't exactly understandable

File: src/store/consts/utils.ts [hunk 5]
```diff
@@ -253,6 +254,27 @@ export const calculateTimeRemaining = (slot: number, nextRoundStartSlot: BN): BN
...
+  synthetics: { [x: string]: ISynthetic },
+  collateral: { [x: string]: ICollateral },
+  nr: number
+) => {
+  let syntheticAddress: string = ''
```
Comment (awojciak) [inline src/store/consts/utils.ts:262]: there wouldn't be a problem when there are multiple syntetics/collaterals with the same asset index? currently we have two collaterals for which this happens

File: src/store/reducers/exchange.ts [hunk 6]
```diff
@@ -218,7 +218,10 @@ const exchangeSlice = createSlice({
       }
       return state
     },
-    setAssetPrice(_state, _action: PayloadAction<{ tokenIndex: number; price: Decimal }>) {},
+    setAssetPrice(
+      _state,
+      _action: PayloadAction<{ tableAddress: string[]; tokenIndex: number; price: Decimal }>
```
Comment (awojciak) [inline src/store/reducers/exchange.ts:223]: tableAddress? there is an array, rename this

File: src/store/sagas/leverage.ts [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+} from '@selectors/exchange'
+import { accounts } from '@selectors/solanaWallet'
+import { getExchangeProgram } from '@web3/programs/exchange'
+import { BN } from '@project-serum/anchor'
+import { spawn, all, select, put, call, takeEvery, SagaGenerator } from 'typed-redux-saga'
+import { DEFAULT_PUBLICKEY } from '@consts/static'
+import { createAccount, getWallet, signAllTransaction, sleep } from './wallet'
+import { currentlySelected } from '@selectors/leverage'
+import { actions as snackbarsActions } from '@reducers/snackbars'
+// import { openLeveragePosition } from './exchange'
```
Comment (PawelKrawczyk96) [inline src/store/sagas/leverage.ts:16]: remove if unnecessary

File: src/store/sagas/leverage.ts [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+    )
+    yield* put(actions.setVaultAddress({ vaultAddress: address.vaultAddress }))
+  } catch (error) {
+    yield* put(actions.setVaultAddress({ vaultAddress: DEFAULT_PUBLICKEY }))
+  }
+}
+
+export function* handleOpenLeverage(): Generator {
+  // yield* call(checkVaultEntry)
+  // yield* call(setVaultAddress)
```
Comment (PawelKrawczyk96) [inline src/store/sagas/leverage.ts:119]: remove this if unnecessary

File: src/store/sagas/leverage.ts [hunk 9]
```diff
@@ -0,0 +1,606 @@
...
+      assetPriceState[vaultCollateral.toString()].val,
+      currentVault.collateralAmount.scale,
+      amountCollateral,
+      (cRatio + 2).toFixed(10)
+    )
+      .mul(new BN(10 ** currentVault.openFee.scale))
+      .div(currentVault.openFee.val.add(new BN(10 ** currentVault.openFee.scale)))
+    console.log('syn price: ', printBN(assetPriceState[vaultSynthetic.toString()].val, 8))
+    console.log('col price: ', printBN(assetPriceState[vaultCollateral.toString()].val, 8))
+    console.log('col price: ', printBN(assetPriceState[vaultCollateral.toString()].val, 8))
```
Comment (PawelKrawczyk96) [inline src/store/sagas/leverage.ts:266]: remove this

**Review by PawelKrawczyk96 — CHANGES_REQUESTED** [review]

File: src/store/consts/utils.ts (hunk 5 above)
Comment (Pawel-Szydlo) [reply src/store/consts/utils.ts:262]: Right, I added the correct condition

File: src/store/selectors/exchange.ts [hunk 10]
```diff
@@ -582,4 +582,19 @@ export const getGeneralTotals = createSelector(
   }
 )
 
+export const getLeverageVaultPairs = createSelector(
+  vaults,
+  synthetics,
+  (allVaults, allSynthetics) => {
+    // eslint-disable-next-line array-callback-return
+    return Object.values(allVaults).filter(vault => {
```
Comment (awojciak) [inline src/store/selectors/exchange.ts:590]: callback in filter should return boolean values, change its body to:
```
() => typeof allSynthetics[vault.synthetic.toString()] !== 'undefined' && typeof allSynthetics[vault.collateral.toString()] !== 'undefined'
```

File: src/store/sagas/wallet.ts [hunk 11]
```diff
@@ -182,6 +182,20 @@ export function* signAndSend(wallet: WalletAdapter, tx: Transaction): SagaGenera
   const signature = yield* call([connection, connection.sendRawTransaction], signedTx.serialize())
   return signature
 }
+export function* signAllTransaction(
```
Comment (awojciak) [inline src/store/sagas/wallet.ts:185]: WalletAdapter already has 'signAllTransactions' method, look at its docs

**Review by awojciak — CHANGES_REQUESTED** [review]
1)
obraz
why is xSOL shown two times?
2) 
obraz

File: src/store/sagas/wallet.ts (hunk 11 above)
Comment (Pawel-Szydlo) [reply src/store/sagas/wallet.ts:185]: ok, but this method not work

File: src/components/Leverage/CloseLeverage/CloseLeverage.tsx [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+            {amount} {tokenTo}
+          </Typography>
+          <Typography className={classes.smallValue}>
+            {percent}% {tokenTo}
+          </Typography>
+        </Grid>
+        <Grid container justifyContent='center' style={{ paddingTop: '16px' }}>
+          <OutlinedButton
+            className={classes.button}
+            name={!blockButton ? 'Close leverage' : "You don't have enouth token to close"}
```
Comment (awojciak) [inline src/components/Leverage/CloseLeverage/CloseLeverage.tsx:122]: *enough

File: src/store/sagas/leverage.ts [hunk 13]
```diff
@@ -0,0 +1,847 @@
...
+            ? 'Insufficient funds for fee'
+            : 'Failed to send. Please try again.',
+        variant: 'error',
+        persist: false
+      })
+    )
+  }
+}
+
+export function* calculateAmountAfterSwap(
```
Comment (awojciak) [inline src/store/sagas/leverage.ts:156]: why is this a saga? it uses no saga effects

**Review by awojciak — CHANGES_REQUESTED** [review]
1) 
obraz
this text looks bad, change it
2) 
obraz
tooltips with explanation why are these assets different from chosen ones would be a good idea

**Review by NorbertBodziony — APPROVED** [review]

**Pull request merged**